File: android/java/im/shimo/bison/gfx/BisonPicture.java
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package im.shimo.bison.gfx;

import android.graphics.Canvas;
import android.graphics.Picture;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;

import java.io.OutputStream;

import im.shimo.bison.CleanupReference;

/**
 * A simple wrapper around a SkPicture, that allows final rendering to be performed using the
 * chromium skia library.
 */
@JNINamespace("bison")
public class BisonPicture extends Picture {
    private long mNativeBisonPicture;

    // There is no explicit destroy method on Picture base-class, so cleanup is always
    // handled via the CleanupReference.
    private static final class DestroyRunnable implements Runnable {
        private long mNativeBisonPicture;
        private DestroyRunnable(long nativeBisonPicture) {
            mNativeBisonPicture = nativeBisonPicture;
        }
        @Override
        public void run() {
            BisonPictureJni.get().destroy(mNativeBisonPicture);
        }
    }

    private CleanupReference mCleanupReference;

    /**
     * @param nativeBisonPicture is an instance of the BisonPicture native class. Ownership is
     *                        taken by this java instance.
     */
    public BisonPicture(long nativeBisonPicture) {
        mNativeBisonPicture = nativeBisonPicture;
        mCleanupReference = new CleanupReference(this, new DestroyRunnable(nativeBisonPicture));
    }

    @Override
    public Canvas beginRecording(int width, int height) {
        unsupportedOperation();
        return null;
    }

    @Override
    public void endRecording() {
        // Intentional no-op. The native picture ended recording prior to java c'tor call.
    }

    @Override
    public int getWidth() {
        return BisonPictureJni.get().getWidth(mNativeBisonPicture, BisonPicture.this);
    }

    @Override
    public int getHeight() {
        return BisonPictureJni.get().getHeight(mNativeBisonPicture, BisonPicture.this);
    }

    @Override
    public void draw(Canvas canvas) {
        BisonPictureJni.get().draw(mNativeBisonPicture, BisonPicture.this, canvas);
    }

    @SuppressWarnings("deprecation")
    public void writeToStream(OutputStream stream) {
        unsupportedOperation();
    }

    private void unsupportedOperation() {
        throw new IllegalStateException("Unsupported in BisonPicture");
    }

    @NativeMethods
    interface Natives {
        void destroy(long nativeBisonPicture);
        int getWidth(long nativeBisonPicture, BisonPicture caller);
        int getHeight(long nativeBisonPicture, BisonPicture caller);
        void draw(long nativeBisonPicture, BisonPicture caller, Canvas canvas);
    }
}

File: android/browser/gfx/bison_picture.cc
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "bison/android/browser/gfx/bison_picture.h"

#include "bison/android/browser/gfx/java_browser_view_renderer_helper.h"
#include "bison/android/browser_jni_headers/BisonPicture_jni.h"
#include "third_party/skia/include/core/SkPicture.h"

using base::android::JavaParamRef;

namespace bison {

BisonPicture::BisonPicture(sk_sp<SkPicture> picture) : picture_(std::move(picture)) {
  DCHECK(picture_);
}

BisonPicture::~BisonPicture() {}

void BisonPicture::Destroy(JNIEnv* env) {
  delete this;
}

jint BisonPicture::GetWidth(JNIEnv* env, const JavaParamRef<jobject>& obj) {
  return picture_->cullRect().roundOut().width();
}

jint BisonPicture::GetHeight(JNIEnv* env, const JavaParamRef<jobject>& obj) {
  return picture_->cullRect().roundOut().height();
}

void BisonPicture::Draw(JNIEnv* env,
                     const JavaParamRef<jobject>& obj,
                     const JavaParamRef<jobject>& canvas) {
  const SkIRect bounds = picture_->cullRect().roundOut();
  std::unique_ptr<SoftwareCanvasHolder> canvas_holder =
      SoftwareCanvasHolder::Create(canvas, gfx::Vector2d(),
                                   gfx::Size(bounds.width(), bounds.height()),
                                   false);
  if (!canvas_holder || !canvas_holder->GetCanvas()) {
    LOG(ERROR) << "Couldn't draw picture";
    return;
  }
  picture_->playback(canvas_holder->GetCanvas());
}

}  // namespace bison
